qemu/tests/sr_iov_lifecycle.py

```diff
@@ -0,0 +1,112 @@
+
+    msg = "Try to boot up guest(s) with VF(s)."
+    error_context.context(msg, logging.info)
+    timeout = int(params.get("login_timeout", 30))
+
+    for vm_name in params["vms"].split(" "):
+        params["start_vm"] = "yes"
+        vm = env.get_vm(vm_name)
+        session = vm.wait_for_serial_login(
+            timeout=int(params.get("login_timeout", 720)))
+        adapter_list = session.get_command_output(
+            "lspci -nn | grep '%s'" % vf_filter)
+        logging.info("Probed VF adapters in guest are:"
+                     " %s", adapter_list)
+
+        if params.get("vm_suspend", "no") == "yes":
+            error_context.base_context("Suspending the VM", logging.info)
+            vm.pause()
+            error_context.context("Verify the status of VM is"
+                                  "'paused'", logging.info)
+            vm.verify_status("paused")
+
+            error_context.context("Verify the session has"
+                                  " no response", logging.info)
+            if session.is_responsive():
+                msg = "Session is still responsive after suspend"
+                logging.error(msg)
+                test.fail(msg)
+            session.close()
+
+            time.sleep(20)
+            error_context.base_context("Resuming the"
+                                       "VM", logging.info)
+            vm.resume()
+            error_context.context("Verify the status of VM is"
+                                  "'running'", logging.info)
+            vm.verify_status("running")
+
+            error_context.context("Re-login the guest", logging.info)
+            session = vm.wait_for_serial_login(
+                timeout=int(params.get("login_timeout", 360)))
+            adapter_list_suspend = session.get_command_output(
+                "lspci -nn | grep '%s'" % vf_filter)
+            logging.info("Adapter details after suspend/resume "
+                         "are %s ", adapter_list_suspend)
+            if adapter_list != adapter_list_suspend:
+                test.fail("Mismatch in adapter list, after suspend/resume.")
+
+            error_context.context("Verify Host and guest kernel "
+                                  "no error and call trace", logging.info)
+            vm.verify_kernel_crash()
+
+        if params.get("vm_reboot", "no") == "yes":
+            error_context.context("Rebooting Guest", logging.info)
+            session.cmd('reboot & exit', timeout=10, ignore_all_errors=True)
+            session = vm.wait_for_serial_login(
+                timeout=int(params.get("login_timeout", 720)))
+            adapter_list_after = session.get_command_output(
+                "lspci -nn | grep '%s'" % vf_filter)
+            logging.info("Adapter details after reboot "
+                         "are %s ", adapter_list_after)
+            if adapter_list != adapter_list_after:
+                test.fail("Mismatch in adapter list, after reboot.")
+        session.close()
```

Comment:
consider having try block for handling interruption during tests and perform cleanup in finally block